### PR TITLE
HPCC-32688 Add JirabotMerge Github Action

### PR DESCRIPTION
**Pull request by jpmcmu**
- Added a Github action to update Jira issues upon PR merger

Signed-off-by: James McMullan James.McMullan@lexisnexis.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [X] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [X] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [X] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the CONTRIBUTORS document.
- [X] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [X] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jpmcmu** [comment]
**Notes on merge script:** The script uses the target branch and the git tags to determine the correct fix versions to add to the Jira issue. The logic for this is as follow:

- If merging into master, find latest tag version and increment to next major. IE: merging while 9.8.X is the latest release will result in a 9.10.0 fix version
- If merging into candidate-9.X, find the latest point in that minor increment it and to fix versions, increment minor and repeat process until we are at latest minor

Here is an example run from my fork merging into master. It added 9.10.0 to the Jira issue: https://github.com/jpmcmu/HPCC-Platform/actions/runs/10951362829

**Review by ghalliday — APPROVED** [review]
Looks good afaics.  I have added the relevant variable.

File: .github/workflows/jirabot-merge.yml [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+      - name: "Install dependencies"
+        run: |
+          set -xe
+          python -VV
+          python -m site
+          python -m pip install --upgrade pip setuptools wheel
+          python -m pip install --upgrade atlassian-python-api
+          python -m pip install --upgrade jira
+      - name: "Checkout"
+        uses: actions/checkout@v2
```
Comment (GordonSmith) [inline .github/workflows/jirabot-merge.yml:36]: Should be version 4

File: .github/workflows/jirabot-merge.yml [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+      - name: "Debug Vars"
+        run: |
+          echo "JIRA_URL: ${{ vars.JIRA_URL }}"
+          echo "Pull Request Number: ${{ github.event.pull_request.number }}"
+          echo "Pull Request Title: ${{ github.event.pull_request.title }}"
+          echo "Pull Request Author Name: ${{ github.event.pull_request.user.login }}"
+          echo "Pull Request URL: ${{ github.event.pull_request.html_url }}"
+          echo "Comments URL: ${{ github.event.pull_request.comments_url }}"
+          echo "Branch Name: ${{ github.ref_name }}"
+      - uses: "actions/setup-python@v2"
```
Comment (GordonSmith) [inline .github/workflows/jirabot-merge.yml:24]: Should be version 5?

File: .github/workflows/jirabot-merge.yml [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+
+                # Escape the result for JSON
+                result = json.dumps(result)
+
+                subprocess.run(['curl', '-X', 'POST', comments_url, '-H', 'Content-Type: application/json', '-H', f'Authorization: token {github_token}', '--data', f'{{ "body": {result} }}'], check=True)
+            else:
+                print('Unable to find Jira issue name in title')
+
+            print(result)
+        shell: python
```
Comment (GordonSmith) [inline .github/workflows/jirabot-merge.yml:291]: Minor: Move shell to above the run for consistency with other actions

**Review by GordonSmith — APPROVED** [review]
Looks good - I didn't review the python code...

**Review by GordonSmith — CHANGES_REQUESTED** [review]
Just minor changes

**Comment by jpmcmu** [comment]
@GordonSmith thanks for the review, made those changes

**Comment by ghalliday** [comment]
@jpmcmu please look at the security warning - which does indicate a real vulnerability

**Comment by jpmcmu** [comment]
@ghalliday Addressed the security issue and squashed

**Review by GordonSmith — APPROVED** [review]

**Comment by ghalliday** [comment]
@jpmcmu can you explain how your change avoids the security problem?  Will it not sill be possible for someone to insert arbitrary text into a title, which will then be echoed?  I haven't understood why putting it in an environment variable first helps (other than confusing the checking tool)

**Comment by jpmcmu** [comment]
@ghalliday Good question, I followed Github's recommendations on how to address the issue, but should have added a comment about why the changes addressed the security issue. (GitHubs: recommendation: https://github.com/hpcc-systems/HPCC-Platform/security/code-scanning/278)

Bash apparently treats the environment variable as a single string which is escaped and prevents injection.

**Pull request merged**